src/Microsoft.AspNet.Mvc.Core/Formatters/AcceptHeaderParser.cs

```diff
@@ -0,0 +1,179 @@
+            if (string.IsNullOrEmpty(value) || (index == value.Length))
+            {
+                return true;
+            }
+
+            var separatorFound = false;
+            var current = GetNextNonEmptyOrWhitespaceIndex(value, index, out separatorFound);
+
+            if (separatorFound && !true)
+            {
+                return false; // leading separators not allowed if we don't support multiple values.
+            }
+
+            if (current == value.Length)
+            {
+                if (true)
+                {
+                    index = current;
+                }
+                return true;
+            }
+
+            MediaTypeSegmentWithQuality result;
+            var length = GetMediaTypeWithQualityLength(value, current, out result);
+
+            if (length == 0)
+            {
+                return false;
+            }
+
+            current = current + length;
+            current = GetNextNonEmptyOrWhitespaceIndex(value, current, out separatorFound);
+
+            // If we support multiple values and we've not reached the end of the string, then we must have a separator.
+            if ((separatorFound && !true) || (!separatorFound && (current < value.Length)))
+            {
+                return false;
+            }
+
+            index = current;
+            parsedValue = result;
+            return true;
+        }
+
+        private static int GetNextNonEmptyOrWhitespaceIndex(
+            string input,
+            int startIndex,
+            out bool separatorFound)
+        {
+            Contract.Requires(input != null);
+            Contract.Requires(startIndex <= input.Length); // it's OK if index == value.Length.
+
+            separatorFound = false;
+            var current = startIndex + HttpTokenParsingRules.GetWhitespaceLength(input, startIndex);
+
+            if ((current == input.Length) || (input[current] != ','))
+            {
+                return current;
+            }
+
+            // If we have a separator, skip the separator and all following whitespaces. If we support
+            // empty values, continue until the current character is neither a separator nor a whitespace.
+            separatorFound = true;
+            current++; // skip delimiter.
+            current = current + HttpTokenParsingRules.GetWhitespaceLength(input, current);
+
+            while ((current < input.Length) && (input[current] == ','))
+            {
+                current++; // skip delimiter.
+                current = current + HttpTokenParsingRules.GetWhitespaceLength(input, current);
+            }
+
+            return current;
+        }
+
+        private static int GetMediaTypeWithQualityLength(
+            string input,
+            int start,
+            out MediaTypeSegmentWithQuality result)
+        {
+            result = default(MediaTypeSegmentWithQuality);
+            var enumerator = new MediaTypeParser(input, start, length: null).GetEnumerator();
```

Comment:
I'm not convinced that this is super valuable as an abstraction vs a `TryGetNextParameter` method